lgc/patch/VertexFetch.cpp

```diff
@@ -1132,6 +1135,31 @@ bool VertexFetchImpl::needSecondVertexFetch(const VertexInputDescription *inputD
   return inputDesc->dfmt == BufDataFormat64_64_64 || inputDesc->dfmt == BufDataFormat64_64_64_64;
 }
 
+// =====================================================================================================================
+// Checks whether the vertex fetch operation needs to be splitted according to the given data format.
+//
+// @param dfmt : Vertex buffer data format
+bool VertexFetchImpl::needSplitFetch(unsigned dfmt) const {
+  bool needSplitFetch = false;
+
+  // NOTE: For these formats, there is a HW defect when the binding offset of vertex buffer is unaligned, we need to
```

Comment:
A HW defect?  Which hardware, and should we be checking for that hardware so that we can still get the better code for other architectures. ?

Reply:
I believe this affects most hardware. As I tested, it occurs at least on gfx8, 9 and 10.

Reply:
gfx8 and gfx9 are unaffected on Linux. You were probably seeing a consequence of the Windows people's boneheaded refusal to enable unaligned accesses.

Reply:
If this is a Windows problem, then can we have different behaviour for Windows and Linux?  Or is that something that is not done?  Splitting the fetch will be be a performance lose.  Also, if the the issue is the OS and not the HW, then the comment should say that.

Reply:
> Also, if the the issue is the OS and not the HW, then the comment should say that.

This is because the kernel mode driver adopts different alignment strategies, which makes this issue an OS-dependent one.

Reply:
So, should I make this an OS dependent behavior, or should I look for some Windows people to enable unaligned accesses?

Reply:
Should we use _WIN32 and _WIN64 macros to gate this piece of codes and provide some explanations as a tentative solution?

Reply:
I would be okay with that, but it is not up to me to make that decision.

Reply:
I don't think this is a good approach, because it also affects the behavior of standalone `amdllpc`, which should be OS-independent.

LLVM has added feature flags for aligned vs. unaligned access recently, maybe the handling can be integrated with those?

Reply:
It would be very helpful if you can point me to the instruction of using the flags.

Thanks in advanced.

Reply:
I think you're looking for `Subtarget->hasUnalignedBufferAccess`

Reply:
Thank you, I'll check it later.

Reply:
@nhaehnle, setting the attribute make no difference, maybe I have some misunderstanding?

What does `the handling can be integrated with those` exactly mean?

Reply:
What I meant is that you apply the workaround in this PR if and only if `!Subtarget->hasUnalignedBufferAccess`.

Reply:
Excuse for my ignorance, I have several more questions to ask.

Are we able to get the subtarget information at this phase? I see nowhere-else in llpc use the subtarget.

And I only see that UnalignedBufferAccess is set only when the OS type is AMDHSA. Will that mean some modification need to be done in llvm instead aside from llpc?

Reply:
I think you can get the `MCSubtargetInfo` from the `TargetMachine`, from the `LgcContext`. But @trenouf may have some other opinion on how to best access this?